### PR TITLE
Enable NFT for ServiceAgreementRegistry

### PR DESCRIPTION
**Pull request by mzxyz**
## Changes

- Enable NFT for service agreement registry
- Remove `establishServiceAgreement` function to simplify the function call for creating agreements
- Split indexer states and functions related from `ServiceAgreementRegistry` to `IndexerServiceAgreement` contract

## Contract Size Changes

- ServiceAgreementRegistry: `26KB` -> `21KB`
- IndexerServiceAgreement : `8KB`

Will fix the test cases in another PR

### CODE REVIEW

File: contracts/IndexerServiceAgreement.sol
Comment (ianhe8x) [inline contracts/IndexerServiceAgreement.sol:1]: Suggest naming it ServiceAgreementRegistryHelper, so it's clear what this is for and align with the naming convention like we did for Rewards contract

File: contracts/ServiceAgreementRegistry.sol [hunk 1]
```diff
@@ -269,50 +175,25 @@ contract ServiceAgreementRegistry is Initializable, OwnableUpgradeable, IService
         _establishServiceAgreement(newAgreementId, false);
     }
 
-    function clearEndedAgreement(address indexer, uint256 id) public {
-        require(id < indexerCsaLength[indexer], 'SA001');
+    function clearEndedAgreement(address indexer, uint256 id) external {
```
Comment (ianhe8x) [inline contracts/ServiceAgreementRegistry.sol:178]: `clearEndedAgreement` here is redundant with `removeEndedAgreement` in the new helper contract

File: contracts/ServiceAgreementRegistry.sol [hunk 2]
```diff
@@ -269,50 +175,25 @@ contract ServiceAgreementRegistry is Initializable, OwnableUpgradeable, IService
...
-        closedServiceAgreementIds[indexer][id] = closedServiceAgreementIds[indexer][indexerCsaLength[indexer] - 1];
-        delete closedServiceAgreementIds[indexer][indexerCsaLength[indexer] - 1];
-        indexerCsaLength[indexer] -= 1;
-        indexerDeploymentCsaLength[indexer][agreement.deploymentId] -= 1;
+        indexerServiceAgreement.removeEndedAgreement(id, agreement);
 
         emit ClosedAgreementRemoved(agreement.consumer, agreement.indexer, agreement.deploymentId, agreementId);
     }
 
-    function clearAllEndedAgreements(address indexer) public {
```
Comment (ianhe8x) [inline contracts/ServiceAgreementRegistry.sol:292]: this is still useful I think, and we use it a lot in tests

**Review by nz-neooo — APPROVED** [review]

**Comment by ianhe8x** [comment]
@mzxyz found an issue with renewAgreement
1. transfer need to happen before `createClosedServiceAgreement`
2. `_establishServiceAgreement` was called twice
3. `_establishServiceAgreement` need to be called with `checkThreshold=false`

**Review by ianhe8x — DISMISSED** [review]
renewAgreement need fix

**Pull request merged**